cranelift-codegen/shared/src/isa/x86/encoding_bits.rs

```diff
@@ -0,0 +1,193 @@
+//! Provides a named interface to the `u16` Encoding bits.
+
+/// Named interface to the `u16` Encoding bits, representing an opcode.
+///
+/// Cranelift requires each recipe to have a single encoding size in bytes.
+/// X86 opcodes are variable length, so we use separate recipes for different
+/// styles of opcodes and prefixes. The opcode format is indicated by the
+/// recipe name prefix.
+///
+/// VEX/XOP and EVEX prefixes are not yet supported.
+/// Encodings using any of these prefixes are represented by separate recipes.
+///
+/// The encoding bits are:
+///
+/// 0-7:   The opcode byte <op>.
+/// 8-9:   pp, mandatory prefix:
+///        00: none (Op*)
+///        01: 66   (Mp*)
+///        10: F3   (Mp*)
+///        11: F2   (Mp*)
+/// 10-11: mm, opcode map:
+///        00: <op>        (Op1/Mp1)
+///        01: 0F <op>     (Op2/Mp2)
+///        10: 0F 38 <op>  (Op3/Mp3)
+///        11: 0F 3A <op>  (Op3/Mp3)
+/// 12-14  rrr, opcode bits for the ModR/M byte for certain opcodes.
+/// 15:    REX.W bit (or VEX.W/E)
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct EncodingBits(u16);
+
+impl From<u16> for EncodingBits {
+    fn from(bits: u16) -> EncodingBits {
+        EncodingBits { 0: bits }
+    }
+}
+
+impl EncodingBits {
+    /// Constructs a new EncodingBits from parts.
+    pub fn new(op_bytes: &[u8], rrr: u16, rex_w: u16) -> Self {
+        let prefix: u16 = OpcodePrefix::from_opcode(op_bytes) as u16;
+        let opcode_byte: u16 = u16::from(op_bytes[op_bytes.len() - 1]);
+
+        debug_assert!(opcode_byte <= 0b1111_1111);
+        debug_assert!(prefix <= 0b1111);
+        debug_assert!(rrr <= 0b111);
+        debug_assert!(rex_w <= 0b1);
```

Comment:
I wonder if these shouldn't be just regular checks that panic even in optimized builds...